### ISSUE TITLE
Callable instances don't complete in the argument list

### ISSUE THREAD

**Issue by bwilkerson** [dart-lang/sdk#45595]
(from https://github.com/dart-lang/sdk/issues/43919#issuecomment-813027803)

Related to this I tried to get around this lack of autocomplete by instead using callable classes, which helps somewhat if you are OK with users having to type '.call' to get autocompletion for your api:

```
class CallableCreate{
  int call(int a){
    return a + 10;
  }
}
CallableCreate create = CallableCreate();

int create2({required int a}) => a + 1;

// Using .call makes nice autocompletion
var x = create.call(10);

// There is no help at all however when I directly call the class :(
var y = create(10);
```

However if you do not start typing .call on the callable class but instead start to directly call it, you get no autocompletion help. If this is also an oversight I am happy to open a new issue for it if you would like? Or perhaps it fits nicely into this issue already.

**Comment by asashour** [comment]
https://dart-review.googlesource.com/c/sdk/+/242283

**Comment by hbock-42** [comment]
Hello, the fix https://dart-review.googlesource.com/c/sdk/+/242283 seems to be in review since more than a year

**Comment by scheglov** [comment]
With https://github.com/dart-lang/sdk/issues/43919 fixed, we do get suggestions for named arguments.
<img width="686" alt="image" src="https://github.com/dart-lang/sdk/assets/384794/b81bd99e-60c7-4443-be42-000b7252c6a3">

However we don't get anything for positional parameters, e.g. when we explicitly invoke `.call()` we get `.call(a)` inserted.
It looks that there were concerns about additional suggestion that implicitly invoked a callable object.
So, an alternative could be to have a suggestion with positional arguments, if the target is an invocable object.
But not if it is normal method - remove it, and re-complete.

@bwilkerson will this be acceptable? My impression that invoking callable objects is rare, and we probably don't get much hit for MRR. OTOH, if it is rare, maybe it is not worth implementing.

**Comment by bwilkerson** [comment]
> So, an alternative could be to have a suggestion with positional arguments, if the target is an invocable object.
> But not if it is normal method - remove it, and re-complete.
> will this be acceptable?

I'm not sure I understand the proposal, so I can't answer that question.

> My impression that invoking callable objects is rare ...

That's my impression too.

> ... and we probably don't get much hit for MRR.

If we're currently not producing any suggestions in those locations, then the MRR can only stay the same or get better.

> OTOH, if it is rare, maybe it is not worth implementing.

I wouldn't say that, but I do think that we should give lower priority to tasks that will have a lower net benefit across all users than we do for tasks that will have a higher net benefit.

**Comment by scheglov** [comment]
We do have suggestions, see
<img width="482" alt="image" src="https://github.com/dart-lang/sdk/assets/384794/1258906f-6ff7-4596-92cf-e3a4e4734d1b">

I propose adding `a, b` to the top, with linked navigation just like completing a method `foo` would insert `foo(a, b)`. Here we already have `foo()`, so insert only argument stubs. Only when the argument list is still empty.

**Comment by bwilkerson** [comment]
I'm not sure how I feel about it, honestly.

I do think there might be some value in giving users a template to fill out for an invocation (though signature help is arguably better), and maybe wanting to add the template after typing the parens is a common enough use case to make it valuable. I wonder, though, whether it might be better to propose the invocation template when completing `foo`.

I also have to wonder how we're going to measure the value of such a completion. It seems unlikely that we're going to find too many places where the argument list consists of a list of variable names where each variable name matches the parameter name of the method, so this is rarely going to match the actual text after the code is complete.

What do you see as the value proposition for this?

**Comment by scheglov** [comment]
We will not able to measure this.
Not without actively observing users.

Proposing the invocation template as in [this CL](https://dart-review.googlesource.com/c/sdk/+/242283)?
I don't see much value here anyway, it is rare case.

**Comment by bwilkerson** [comment]
> Proposing the invocation template as in [this CL](https://dart-review.googlesource.com/c/sdk/+/242283)?

Yes, though I don't disagree with my earlier concerns and would want both the performance and the MMR impact to be measured.

**Comment by scheglov** [comment]
I did some measurements, count the number of `FunctionExpressionInvocation` when the `node.function.staticType is InterfaceType`.
```
analyzer:
[countFunctionExpressionInvocation: 18]
[countMethodInvocation: 117098]
[countInstanceCreationExpression: 9454]


flutter:
[countFunctionExpressionInvocation: 0]
[countMethodInvocation: 390090]
[countInstanceCreationExpression: 225429]
```

So, it looks to me that the situation described by this issue almost does not exist.
So, fixing it will not add any value.
So, I propose closing it as `WillNotFix`.

**Comment by bwilkerson** [comment]
The evidence you provided suggests that this isn't a style used very often by the Dart and Flutter teams, but (a) we have no evidence that those teams are representative of our larger user base and (b) we have evidence that this bug is impacting at least one user. While it's reasonable to question the priority of this issue, I don't believe that it's reasonable to say that we won't fix this bug.

**Comment by scheglov** [comment]
You previously indicated that we need to measure MMR impact, what code base do you propose using for this?

I want also say that I find it hard to argue with "we have no evidence that XYZ are representative". Do you have an idea of a possible evidence for a set of code bases that will let us say that we will consider them representative?

**Comment by hbock-42** [comment]
If I can add up about the measurement, I would argue that people not using it might be related to this issue.
I cannot speak for others, but I personally don't use this feature BECAUSE the autocomplete does not work and, like a lot of developers, I am lazy. I am also thinking about people working with me and don't want to use a feature that might make their work less easy (I agree the word is a bit strong).
Statistics are useful but context is very important.

**Comment by scheglov** [comment]
I think being lazy is a perfect reason :-)

I tried to use callable classes myself on couple occasions, but it did not stick. If I need to pass a function, I can use a method tear-off. If something accepts an interface, invoke one of its methods. It feels that by doing `myClass()` we omit a useful signal - the name of the operation, the name of the method.

I also buy the argument that it is not used because it is inconvenient only to some extent - there are people who use unsophisticated editors, or less lazy. And if it were useful, we would see more uses.

That said, I agree with @bwilkerson that the analyzer and flutter alone are not representative. And if there are open source projects that are good in some sense: continue to have updates, successfully evolve, large enough; I'd like to include them into the set to measure metrics.

**Comment by scheglov** [comment]
OK, apparently my involvement is not useful here.
I will let someone else drive this.
